### ISSUE TITLE
Support global config for name casing (table, fields)

### ISSUE THREAD

**Issue by sandbox-apps** [prisma/prisma#8283]
## Problem
Mapping database table name and field name is tiring and counter productive both (creating new project or coming from existing database project)

## Suggested solution
generator client {
  provider = "prisma-client-js"
  casing = "snake-case" // camel-case, pascal-case, kebab-case
}

Instead of mapping every table name and column name with @map @@map
we can instead use global config like that

**Comment by janpio** [comment]
What exactly would happen on a schema and a configuration of `casing = "snake-case"` on the database side? Would be nice to illustrate that and all the other cases with examples so it is clear what we are talking about here. 



> ```graphql
>   pluralize = true
> ```
> ...
> `pluralize` also included to map model name to db ex: from `model User{}` to becoming `users` in database

That sounds like a related, but different feature request that I would ask you to please turn into its own issue and also provide an example how that would look like in practice. Thanks.

**Comment by sandbox-apps** [comment]
For the casing side:
Inside `schema.prisma` we have

model User {
  firstName String
  lastName  String
}
model Post {}
model Like {}

Current Output on the database table names:
1. User
2. Post
3. Like
4. UserFollower

If we apply `casing = "snake-case"` the output for table names will be
1. user
2. post
3. like
4. user_follower

this would also propagate to column names: `casing = "snake-case"`
1. first_name
2. last_name

The `pluralize = true` functionality only apply to the model/table name
1. users
2. posts
3. likes
4. user_followers
5. if we have `model Category {}` it will be translated to `categories` in the database

**Comment by pantharshit00** [comment]
Related to https://github.com/prisma/prisma/issues/2790

**Comment by nikolasburk** [comment]
For everyone who is interested in this, check out this tool as a workaround that can be used to automatically transform snake_case to camelCase: https://github.com/IBM/prisma-schema-transformer

**Comment by HairyRabbit** [comment]
any update?

**Comment by AradAlvand** [comment]
Hi there, I'm new to Prisma (coming from an EF Core background), and it's disappointing to see that Prisma doesn't have a feature like this. Having to manually writing mapping for table/field names (when using Postgres where the convention is all-lowercase snake-case), when it's a fairly easy-to-automate task, is a HUGE bummer.

Hope you give this issue a little more love.

**Comment by rclmenezes** [comment]
If anyone wants a temporary solution, here's mine:
https://github.com/prisma/prisma/issues/1934#issuecomment-1304805678

**Comment by janpio** [comment]
Yes, we definitely plan to work on this one day. Right now we are still collecting input.

So: Can you elaborate on the MikroORM mention here? What exactly do they do how?

**Comment by maximeburri** [comment]
Nice.

During the initialization you can define the naming strategy with `namingStrategy: MyCustomNamingStrategy` (https://mikro-orm.io/docs/naming-strategy). You have some default ones:
- [The underscore naming strategy](https://github.com/mikro-orm/mikro-orm/blob/4d2ea6aee19fbd1d2765ccd38ca5c75b690a83a0/packages/core/src/naming-strategy/MongoNamingStrategy.ts) (`snake-case`) basically do a `message.replace(/([a-z])([A-Z])/g, '$1_$2').toLowerCase()` 
- [Mongo naming strategy](https://github.com/mikro-orm/mikro-orm/blob/4d2ea6aee19fbd1d2765ccd38ca5c75b690a83a0/packages/core/src/naming-strategy/UnderscoreNamingStrategy.ts) (`kebab-case`) do a `replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase()`

That's just an implementation, but I like the way of being free to implement its own strategy by inheriting from the abstract class if the default ones are not enough.

And of course `@Map` and `@@Map` would have precedence.

**Comment by janpio** [comment]
Ok, so you basically define the strategy and they then apply that to your "entity" name to create the table and field names?
So in Prisma you would want to be able to set that, and then e.g. `model UserTest` would automatically become a table `user_test` with the respective naming strategy, without having to manually use `@@map` on the model?

**Comment by maximeburri** [comment]
Exactly that would be perfect.

For example:
```prisma
model SuperUser {
  id        Int      @id @default(autoincrement())
  firstName String
  @map("map_last_name")
  lastName  String
  email     String
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt
}
```

Would create:
- For the  `snake-case`: table `super_user` with fields `id`, `first_name`, `map_first_name`, `email`, `created_at`, `updated_at`
- For the  `kebab-case`: table `super-user` with fields `id`, `first-name`, `map_first_name`, `email`, `created-at`, `updated-at`
- For the  `camel-case` (the actual case by default): table `SuperUser` with fields `id`, `firstName`, `map_first_name`, `email`, `createdAt`, `updatedAt`

Note that the `@map` takes precedence (same for the `@@map`)

**Comment by aniravi24** [comment]
Currently using https://github.com/botika/prisma-schema-transformer to workaround this, but it uses the DMMF API which is internal and could break at any time. Requires maintaining a lot of code for something Prisma probably could do more easily within its internal codebase. Whenever new Prisma features are announced, this has to be updated sometimes to make sure the schema still prints correctly, even if the code changes have [nothing to do with casing](https://github.com/botika/prisma-schema-transformer/pull/173).

Similar to what people said above, I like to maintain snake case within a database like PostgreSQL, but want application code in TypeScript to be camel case to match the convention there, and having to manually add the mapping everywhere can be tedious.

**Comment by alexmantaut** [comment]
Adding another one to the list of workarounds...

This tool (https://github.com/iiian/prisma-case-format) allows to check your schema.prisma for consistency with your naming conventions...

You can execute it whenever you change your model files, and it will generate all the adequate `@map()` and `@@map ` annotations.

In my case I wanted to have camelCase attributes on TS and snake_case on the DB, so I did:

So I executed this:
```
npx prisma-case-format --map-table-case snake --map-field-case snake
```

Not sure if it covers all use cases, but I hope it helps

**Comment by sjohnson540** [comment]
I would love to take a crack at implementing this if that is something the maintainers are interested in. I think having the global config makes sense. Although I agree with other commenters that it really makes more sense to translate by default into what the database uses. I.e. postgres is meant to use snake_case (although you can get around it by quoting obviously).

Anyway, I have some free time and this really bugs me so would love to give it a shot.

**Comment by markoorn** [comment]
@samcraigjohnson - any progress on this? ;) If not I might have to beat you to it

**Comment by samcraigjohnson** [comment]
> @samcraigjohnson - any progress on this? ;) If not I might have to beat you to it

@markoorn go for it, I was told on twitter that the suggested method would be to create my own library to do it, so my impression was they weren't planning on merging it even if I did the work.

**Comment by p-runge** [comment]
@samcraigjohnson Could you link the tweet please? I'd be interested in why they made that decision. Sounds like a strange statement regarding a common issue like this one.

**Comment by mrbrannan** [comment]
> @samcraigjohnson Could you link the tweet please? I'd be interested in why they made that decision. Sounds like a strange statement regarding a common issue like this one.

@p-runge https://twitter.com/nikolasburk/status/1714915567558128072

**Comment by p-runge** [comment]
> > @samcraigjohnson Could you link the tweet please? I'd be interested in why they made that decision. Sounds like a strange statement regarding a common issue like this one.
> 
> @p-runge https://twitter.com/nikolasburk/status/1714915567558128072

Thanks!

@samcraigjohnson
My impression of this is that @nikolasburk  just offered you doing an own lib since that would be faster than Prisma to accept the PR. But the PR being a totally valid option as well that might just take a couple of iterations due to edge cases not being considered.

So in case you actually updated your own fork of Prisma, I think it might still be worth it to make a PR out of it.

**Comment by Shahid-Habibi714** [comment]
Totally agree that having a one-line solution for casing would save a lot of headaches.

Regarding `pluralize: true`, Prisma could automatically append an "s" to each model name by default. For cases where this rule doesn't work, we could define an explicit override using a decorator-like syntax, for example:

```prisma
@pluralize("Categories")
model Category { ... }
```

This would be similar in spirit to how decorators are used in NestJS, while still allowing flexibility for irregular plural forms.

**Comment by jacknkandy** [comment]
Would love this functionality. Just moving a database from Supabase into a dedicated Postgres database. The code base is primarily using snake case and fact that Prisma is forcing camel case has made refactoring very difficult and confusing.

**Comment by mordechaim** [comment]
[OrchidORM](https://github.com/romeerez/orchid-orm) has built in snake to camel support. It's a far more superior ORM in my opinion.
